Skip unused volumes in VolumeManager

jsafrane opened the pull request:
DesiredStateOfWorldPopulator should skip a volume that is not used in any pod. "Used" means either mounted (via volumeMounts) or used as raw block device (via volumeDevices) in a container in the pod.

Especially when block feature is disabled, a block volume must not get into DesiredStateOfWorld, because it would be formatted and mounted there, potentially overwriting any existing raw block data.

/kind bug
Fixes #76044

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Volumes specified in a pod but not used in it are no longer unnecessarily formatted, mounted and reported in `node.status.volumesInUse`.
```


**Change of Kubernetes behavior**
Previously, all volumes were reported in `node.status.volumesInUse` and they were mounted / mapped as raw block devices in kubelet. With this PR, only the volumes that are actually used will be reported there and mounted/mapped. In most cases this won't make a difference (why would anyone put a volume to a pod and then not use it in any container?)

cc: @wongma7 @msau42 @bertinatto

wongma7 commented:
My old attempt to fix this looks extremely convoluted... What happens in kubelet-side WaitForAttachAndMount since it also reads spec.volumes? https://github.com/kubernetes/kubernetes/blob/b409073e99695ea35642a8194b9285ac12fd0cf8/pkg/kubelet/volumemanager/volume_manager.go#L440

mattjmcnaughton commented:
I believe that the `pull-kubernetes-bazel-test` failure is legit:

```
 --- FAIL: TestGetMountedVolumesForPodAndGetVolumesInUse (123.00s)
    volume_manager_test.go:80: Expected success: unmounted volumes=[vol1], unattached volumes=[vol1]: timed out waiting for the condition 
```

My guess is that the other failures are flake... retesting them now to verify.

/test pull-kubernetes-integration
/test pull-kubernetes-kubemark-e2e-gce-big
/test pull-kubernetes-kubemark-e2e-gce-csi-serial

wongma7 commented:
Yes, test failure is legit, @jsafrane volumemanager needs some kind of fix as well since it gets list of volumes to wait for from spec.volumes instead of desired state. I'm sure you can fix it in a nicer way than I did.

msau42 commented:
/assign @jingxu97

jsafrane commented:
> volumemanager needs some kind of fix as well since it gets list of volumes to wait for from spec.volumes instead of desired state. I'm sure you can fix it in a nicer way than I did.

You're right, I reworked `getExpectedVolumes` a bit. I used the simplest solution I could think of, not sure it's nicer.

In addition, I moved most of the code to `processPodVolumes()`, the code got a bit cleaner, but then  I needed to pass fake `BlockVolumePathHandler` to `NewVolumeManager` for unit tests, so the PR grew a lot...

/hold
* Missing e2e test with unused volume
* Squash the commits

jsafrane commented:
@wongma7, can you please take a look at the current status? I'll squash the commits if this is the way to go.

I could not find any elegant way how to e2e test this. From storage test I can't get the global mount directory or name to check in `node.status.volumesInUse` easily. I *could* compare `volumesInUse` before and after a test pod was created, but that 1) would imply `[Serial]`, 2) it's quite ugly and 3) I am not sure it would work for all plugins. Any ideas?

wongma7 commented:
IMO it doesn't need to work for all plugins. Maybe just gcepd, e.g. https://github.com/kubernetes/kubernetes/blob/1addcf471519dde12667832980201785f6c2628a/test/e2e/storage/pd.go#L616 .

Initially I was thinking to ssh into the node and parse the mounts. Then you could e.g. make sure that the test pod UID has 0 mounts.

wongma7 commented on pkg/kubelet/volumemanager/volume_manager.go:
```diff
@@ -435,13 +438,24 @@ func filterUnmountedVolumes(mountedVolumes sets.String, expectedVolumes []string
 
-	for _, podVolume := range pod.Spec.Volumes {
-		expectedVolumes = append(expectedVolumes, podVolume.Name)
+	for _, container := range pod.Spec.Containers {
```
what about initContainers?

jsafrane replied:
Good catch!

I copied the code from VolumeManager and it observes only `.containers`, not `.initContainers`. Filed #81343. I think the best would be to fix `makeVolumeMap` in #81343 first (to have a small patch), backport it everywhere and then move it to `pkg/volume/utils` that will be used also here.

jsafrane replied:
Wrote too early, the issue is reproducible only in this PR.

I extracted `GetPodVolumeNames` and used it consistently in VolumeManager and DSWP.

wongma7 commented on pkg/kubelet/volumemanager/volume_manager_test.go:
```diff
@@ -51,45 +54,103 @@ const (
+			pvMode:      v1.PersistentVolumeBlock,
+			podMode:     v1.PersistentVolumeFilesystem,
+			expectMount: false,
+			expectError: true,
```
where does the error come from, is it a timeout or won't `expectedVolumes` have len 0?

jsafrane replied:
It's not timeout, DSWP has stored error into DSW and `WaitForAttachAndMount` asynchronously picked it up within ~`podAttachAndMountRetryInterval` (300ms)

jsafrane commented:
> Initially I was thinking to ssh into the node and parse the mounts. Then you could e.g. make sure that the test pod UID has 0 (non secret/token etc.)* mounts.

Ha, we mount unused volumes even to pod directory. I can test that.

jsafrane committed:
Add test for unused volumes

jsafrane committed:
Skip unused volumes in VolumeManager

DesiredStateOfWorldPopulator should skip a volume that is not used in any
pod. "Used" means either mounted (via volumeMounts) or used as raw block
device (via volumeDevices).

Especially when block feature is disabled, a block volume must not get into
DesiredStateOfWorld, because it would be formatted and mounted there.

jsafrane committed:
Refactor makeMountsMap into GetPodVolumeNames

The function will be handy in subsequent patches. Also change custom maps
into sets.String.

jsafrane commented:
Added e2e test. Now it should be complete.

jsafrane commented:
e2e test added, commits squashed a bit
/hold cancel

jsafrane commented:
/retest

msau42 commented on pkg/volume/util/util.go:
```diff
@@ -548,3 +548,31 @@ func IsLocalEphemeralVolume(volume v1.Volume) bool {
+	mounts = sets.NewString()
+	devices = sets.NewString()
+
+	addContainerVolumes(pod.Spec.Containers, mounts, devices)
```
This may need to be updated again after https://github.com/kubernetes/kubernetes/pull/59484

wongma7 commented:
/lgtm
beautiful!

jsafrane commented:
/assign @tallclair @derekwaynecarr 
for approval from node point of view

Note that there is a small behavior change - before this PR, kubelet mounted / mapped as raw block everything that was in `pod.spec.volumes`, now it mounts / maps only things that are in `pod.spec.containers[*].volumeMounts` / `volumeDevices`.

derekwaynecarr commented:
kubelet changes lgtm.

/lgtm
/approve
